File: pcbstatorgen/magnetic/coil_model.py
"""
pcbstatorgen.magnetic.coil_model
=======================================
Converts :class:`~pcbstatorgen.geometry.wave_winding.PhaseCoil` objects
into Magpylib v5 ``current.Polyline`` sources for magnetic field and force
calculations.

Conductor modelling
-------------------
Each *active conductor* segment (the vertical traces that run perpendicular to
the travel axis and actually generate thrust) is represented by a single
``Polyline`` object with two vertices — one segment from start to end.

End-turn segments (horizontal, parallel to travel) are **not included** by
default.  They contribute a Z-direction normal force and a small Y lateral
force, but their net X thrust contribution cancels out across the full coil
and they would roughly double the computation time.  Pass
``include_end_turns=True`` to include them when calculating normal or lateral
forces.

Current sign convention
-----------------------
The ``current`` parameter of a ``Polyline`` is always **positive**.  The
*direction* of current flow is encoded in the vertex order:

* Even-indexed active conductors in the serpentine go from ``(x, 0, 0)`` to
  ``(x, W, 0)``  — current flows in **+Y**.
* Odd-indexed active conductors go from ``(x, W, 0)`` to ``(x, 0, 0)``
  — current flows in **−Y**.

Combined with the alternating pole arrangement (Bz alternates sign with every
pole pitch), both conductor types produce thrust in the same **+X** direction
when the current magnitude is positive and the commutation is correct.

Layer Z offset
--------------
Conductors are placed in the Z = 0 plane by default (PCB top surface).  Pass
``layer_z_m`` to shift them to a specific layer depth when modelling a
multi-layer stackup.  The copper thickness (35–140 µm) is negligible compared
to the air gap (≥ 0.5 mm), so placing all conductors at their layer centre is
adequate for force estimation.
"""

from __future__ import annotations

import magpylib as magpy
import numpy as np

from pcbstatorgen.geometry.wave_winding import CoilSegment, PhaseCoil

__all__ = ["CoilCurrentModel", "PhaseCurrentSources"]

#: Default meshing density for Polyline targets in getFT.
#: Each active segment (~20 mm long) is divided into this many sub-segments
#: for the Lorentz-force integration.  20 gives sub-mm resolution.
DEFAULT_MESHING: int = 20


class PhaseCurrentSources:
    """Magpylib Polyline sources for one phase at one current level.

    Attributes
    ----------
    phase_idx:
        Phase index (0 = A, 1 = B, 2 = C).
    current_a:
        Signed phase current applied to this set of conductors [A].
    polylines:
        List of ``magpy.current.Polyline`` objects, one per included segment.
    """

    def __init__(
        self,
        phase_idx: int,
        current_a: float,
        polylines: list,
    ) -> None:
        self.phase_idx = phase_idx
        self.current_a = current_a
        self.polylines = polylines

    def __len__(self) -> int:
        return len(self.polylines)

    def as_collection(self) -> magpy.Collection:
        """Return all polylines as a single ``magpy.Collection``."""
        return magpy.Collection(*self.polylines)


class CoilCurrentModel:
    """Convert :class:`~pcbstatorgen.geometry.wave_winding.PhaseCoil`
    objects into Magpylib ``current.Polyline`` sources.

    Parameters
    ----------
    meshing:
        Number of sub-segments per conductor for ``getFT`` force integration.
        Higher values are more accurate but slower.  Default: 20.
    include_end_turns:
        If ``True``, end-turn segments are also converted to Polylines.
        Default: ``False`` (active conductors only).
    layer_z_m:
        Z coordinate of the conductor plane [m].  0 = PCB top surface.
        Useful when modelling buried layers in a multi-layer stackup.

    Examples
    --------
    Single-phase sources::

        model = CoilCurrentModel()
        sources = model.build_phase(coil_a, current_a=1.0)
        F, T = magpy.getFT(magnet_coll, sources.polylines)

    All phases at once::

        all_sources = model.build_all_phases(coils, currents=[1.0, -0.5, -0.5])
        flat = [p for src in all_sources for p in src.polylines]
        F, T = magpy.getFT(magnet_coll, flat)
    """

    def __init__(
        self,
        meshing: int = DEFAULT_MESHING,
        include_end_turns: bool = False,
        layer_z_m: float = 0.0,
    ) -> None:
        if meshing < 1:
            raise ValueError(f"meshing must be ≥ 1, got {meshing}")
        self.meshing = meshing
        self.include_end_turns = include_end_turns
        self.layer_z_m = layer_z_m

    # ------------------------------------------------------------------
    # Public builders
    # ------------------------------------------------------------------

    def build_phase(
        self,
        coil: PhaseCoil,
        current_a: float,
    ) -> PhaseCurrentSources:
        """Build Polyline sources for a single phase coil.

        Parameters
        ----------
        coil:
            Phase coil from :class:`~pcbstatorgen.geometry.wave_winding.WaveWindingGenerator`.
        current_a:
            Phase current to apply [A].  Positive = rated peak current.
            The serpentine vertex ordering handles the alternating conductor
            direction automatically.

        Returns
        -------
        PhaseCurrentSources
        """
        segments = coil.active_segments
        if self.include_end_turns:
            segments = coil.segments  # type: ignore[assignment]

        polylines = [
            self._segment_to_polyline(seg, current_a)
            for seg in segments
        ]
        return PhaseCurrentSources(
            phase_idx=coil.phase_idx,
            current_a=current_a,
            polylines=polylines,
        )

    def build_all_phases(
        self,
        coils: list[PhaseCoil],
        currents_a: list[float],
    ) -> list[PhaseCurrentSources]:
        """Build sources for all phases simultaneously.

        Parameters
        ----------
        coils:
            Phase coils in phase order (A, B, C, …).
        currents_a:
            Signed phase current for each coil [A].  Must have the same
            length as ``coils``.

        Returns
        -------
        list[PhaseCurrentSources]
            One entry per phase.

        Raises
        ------
        ValueError
            If ``coils`` and ``currents_a`` have different lengths.
        """
        if len(coils) != len(currents_a):
            raise ValueError(
                f"coils ({len(coils)}) and currents_a ({len(currents_a)}) "
                "must have the same length"
            )
        return [
            self.build_phase(coil, I)
            for coil, I in zip(coils, currents_a)
        ]

    def flat_polylines(
        self,
        sources: list[PhaseCurrentSources],
    ) -> list:
        """Flatten all Polylines from multiple phases into one list.

        Parameters
        ----------
        sources:
            Output of :meth:`build_all_phases`.

        Returns
        -------
        list[magpy.current.Polyline]
            Flat list of all Polylines across all phases — pass directly to
            ``magpy.getFT(magnets, flat_polylines)``.
        """
        return [p for src in sources for p in src.polylines]

    # ------------------------------------------------------------------
    # Field sampling helpers
    # ------------------------------------------------------------------

    def bfield_at_conductor_positions(
        self,
        coil: PhaseCoil,
        magnet_collection: magpy.Collection,
    ) -> np.ndarray:
        """Sample the magnet B field at each active conductor midpoint.

        Useful for debugging the coupling and verifying pole alignment.

        Parameters
        ----------
        coil:
            Phase coil whose active conductor midpoints are the observers.
        magnet_collection:
            Magnet array at the desired fader position.

        Returns
        -------
        np.ndarray
            Shape ``(n_active_conductors, 3)`` — B [T] at each conductor.
        """
        midpoints = np.array([
            (seg.start[0], (seg.start[1] + seg.end[1]) / 2.0, self.layer_z_m)
            for seg in coil.active_segments
        ])
        return magpy.getB(magnet_collection, midpoints)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _segment_to_polyline(
        self,
        seg: CoilSegment,
        current_a: float,
    ) -> magpy.current.Polyline:
        """Convert one CoilSegment to a Magpylib Polyline."""
        return magpy.current.Polyline(
            current=current_a,
            vertices=[
                (seg.start[0], seg.start[1], self.layer_z_m),
                (seg.end[0],   seg.end[1],   self.layer_z_m),
            ],
            meshing=self.meshing,
        )

File: crates/pcbstatorgen-rs/src/magnetic/coil_model.rs
//! `CoilCurrentModel` — converts `PhaseCoil` objects into sampled conductor
//! segments for Lorentz force integration.
//!
//! Ports `pcbstatorgen/magnetic/coil_model.py`.
//!
//! ## Conductor modelling
//!
//! Each *active conductor* segment (vertical traces perpendicular to travel)
//! is subdivided into `meshing` sub-segments for the Lorentz-force integration.
//! End-turn segments are excluded by default (their net X thrust cancels out
//! and they roughly double computation time).
//!
//! ## Current sign convention
//!
//! The current sign is applied directly to the Lorentz force integral
//! `F = I · Σ(dLᵢ × Bᵢ)`. The serpentine vertex ordering handles the
//! alternating conductor direction automatically:
//! - Even-indexed active conductors go from `(x, 0)` to `(x, W)` — current +Y
//! - Odd-indexed go from `(x, W)` to `(x, 0)` — current -Y

use crate::geometry::PhaseCoil;

/// Default meshing density for sub-segment force integration.
/// Each active segment (~20 mm long) is divided into this many sub-segments.
/// 20 gives sub-mm resolution.
pub const DEFAULT_MESHING: usize = 20;

/// A sub-segment sample point for Lorentz force integration.
///
/// `midpoint_3d`: 3D midpoint of the sub-segment [m]
/// `dl_3d`: direction vector × sub-segment length [m] (the `dL` in `dL × B`)
#[derive(Debug, Clone, Copy)]
pub struct ConductorSample {
    /// 3D midpoint of this sub-segment [m]
    pub midpoint_3d: [f64; 3],
    /// Direction × length of sub-segment [m] — the dL vector
    pub dl_3d: [f64; 3],
}

/// Converts `PhaseCoil` objects into sampled conductor sub-segments for
/// Lorentz force integration.
///
/// This is the Rust equivalent of Python's `CoilCurrentModel`. Instead of
/// building `magpylib.current.Polyline` objects, it directly produces
/// `ConductorSample` records (midpoint + dL) that the `ForceEvaluator`
/// feeds into `F = I · Σ(dLᵢ × Bᵢ)`.
#[derive(Debug, Clone)]
pub struct CoilCurrentModel {
    /// Number of sub-segments per conductor for force integration.
    pub meshing: usize,
    /// If `true`, end-turn segments are also converted to samples.
    pub include_end_turns: bool,
    /// Z coordinate of the conductor plane [m]. 0 = PCB top surface.
    pub layer_z_m: f64,
}

impl Default for CoilCurrentModel {
    fn default() -> Self {
        Self {
            meshing: DEFAULT_MESHING,
            include_end_turns: false,
            layer_z_m: 0.0,
        }
    }
}

impl CoilCurrentModel {
    /// Create a new `CoilCurrentModel`.
    ///
    /// # Panics
    /// Panics if `meshing < 1`.
    pub fn new(meshing: usize, include_end_turns: bool, layer_z_m: f64) -> Self {
        assert!(meshing >= 1, "meshing must be >= 1, got {meshing}");
        Self {
            meshing,
            include_end_turns,
            layer_z_m,
        }
    }

    /// Build conductor samples for a single phase coil.
    ///
    /// Returns a `Vec<ConductorSample>` with one entry per sub-segment.
    /// The caller applies the current multiplier when computing force.
    pub fn build_phase_samples(&self, coil: &PhaseCoil) -> Vec<ConductorSample> {
        let segments: Vec<_> = if self.include_end_turns {
            coil.segments.iter().collect()
        } else {
            coil.active_segments().into_iter().collect()
        };

        let mut samples = Vec::with_capacity(segments.len() * self.meshing);
        for seg in segments {
            self.segment_to_samples(seg.start, seg.end, &mut samples);
        }
        samples
    }

    /// Build conductor samples for all phases, with a current per phase.
    ///
    /// Returns `(phase_samples, phase_currents)` where `phase_samples` is
    /// a vec of `(phase_idx, Vec<ConductorSample>)` and `phase_currents`
    /// is the signed current per phase [A].
    pub fn build_all_phases_samples<'a>(
        &self,
        coils: &'a [PhaseCoil],
        currents_a: &[f64],
    ) -> Vec<(&'a PhaseCoil, f64, Vec<ConductorSample>)> {
        assert_eq!(
            coils.len(),
            currents_a.len(),
            "coils ({}) and currents_a ({}) must have the same length",
            coils.len(),
            currents_a.len()
        );
        coils
            .iter()
            .zip(currents_a.iter())
            .map(|(coil, &current)| (coil, current, self.build_phase_samples(coil)))
            .collect()
    }

    /// Convert one segment to sub-segment samples (midpoint + dL).
    ///
    /// Appends `meshing` samples to `out`.
    fn segment_to_samples(
        &self,
        start: (f64, f64),
        end: (f64, f64),
        out: &mut Vec<ConductorSample>,
    ) {
        let n = self.meshing as f64;
        let dx_total = end.0 - start.0;
        let dy_total = end.1 - start.1;
        // Sub-segment dL (full 3D, with z=0 in the conductor plane)
        let dl_x = dx_total / n;
        let dl_y = dy_total / n;
        let dl_z = 0.0;

        for i in 0..self.meshing {
            let t_mid = (i as f64 + 0.5) / n;
            let mx = start.0 + t_mid * dx_total;
            let my = start.1 + t_mid * dy_total;
            out.push(ConductorSample {
                midpoint_3d: [mx, my, self.layer_z_m],
                dl_3d: [dl_x, dl_y, dl_z],
            });
        }
    }

    /// Sample B-field at each active conductor midpoint (for debugging).
    ///
    /// Returns `Vec<[f64; 3]>` of B vectors [T], one per active segment.
    pub fn bfield_at_conductor_midpoints(
        &self,
        coil: &PhaseCoil,
        b_at: impl Fn(&[[f64; 3]]) -> Vec<[f64; 3]>,
    ) -> Vec<[f64; 3]> {
        let midpoints: Vec<[f64; 3]> = coil
            .active_segments()
            .iter()
            .map(|seg| {
                let mid = seg.midpoint();
                [mid.0, mid.1, self.layer_z_m]
            })
            .collect();
        b_at(&midpoints)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::CoilTopology;
    use crate::geometry::{CoilSegment, PhaseCoil};

    fn make_test_coil() -> PhaseCoil {
        PhaseCoil {
            phase_idx: 0,
            layer_idx: 0,
            phase_name: "A".to_string(),
            topology: CoilTopology::Serpentine,
            layer_pair: None,
            center_via_positions: vec![],
            segments: vec![
                CoilSegment { start: (0.0, 0.0), end: (0.0, 0.020), is_active: true },
                CoilSegment { start: (0.0, 0.020), end: (0.012, 0.020), is_active: false },
                CoilSegment { start: (0.012, 0.020), end: (0.012, 0.0), is_active: true },
            ],
        }
    }

    #[test]
    fn test_build_phase_samples_active_only() {
        let model = CoilCurrentModel::new(5, false, 0.0);
        let coil = make_test_coil();
        let samples = model.build_phase_samples(&coil);
        // 2 active segments × 5 sub-segments = 10
        assert_eq!(samples.len(), 10);
    }

    #[test]
    fn test_build_phase_samples_with_end_turns() {
        let model = CoilCurrentModel::new(5, true, 0.0);
        let coil = make_test_coil();
        let samples = model.build_phase_samples(&coil);
        // 3 segments × 5 sub-segments = 15
        assert_eq!(samples.len(), 15);
    }

    #[test]
    fn test_subsegment_midpoints_and_dl() {
        let model = CoilCurrentModel::new(2, false, 0.001);
        let coil = make_test_coil();
        let samples = model.build_phase_samples(&coil);
        // First active segment: (0,0) → (0, 0.020), meshing=2
        // Sub 0 midpoint: (0, 0.005, 0.001), dl: (0, 0.010, 0)
        // Sub 1 midpoint: (0, 0.015, 0.001), dl: (0, 0.010, 0)
        assert_eq!(samples[0].midpoint_3d, [0.0, 0.005, 0.001]);
        assert_eq!(samples[0].dl_3d, [0.0, 0.010, 0.0]);
        assert_eq!(samples[1].midpoint_3d, [0.0, 0.015, 0.001]);
        assert_eq!(samples[1].dl_3d, [0.0, 0.010, 0.0]);
    }

    #[test]
    #[should_panic(expected = "meshing must be >= 1")]
    fn test_meshing_zero_panics() {
        CoilCurrentModel::new(0, false, 0.0);
    }
}
